
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      
        <link rel="prev" href="../model/">
      
      
        <link rel="next" href="../plotting/">
      
      
      <link rel="icon" href="../../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.6.5">
    
    
      
        <title>Utility - CONCORD - Contrastive Learning for Cross-domain Reconciliation and Discovery</title>
      
    
    
      <link rel="stylesheet" href="../../assets/stylesheets/main.8608ea7d.min.css">
      
        
        <link rel="stylesheet" href="../../assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../../assets/_mkdocstrings.css">
    
    <script>__md_scope=new URL("../..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
      
        <meta  property="og:type"  content="website" >
      
        <meta  property="og:title"  content="Utility - CONCORD - Contrastive Learning for Cross-domain Reconciliation and Discovery" >
      
        <meta  property="og:description"  content="None" >
      
        <meta  property="og:image"  content="./assets/images/social/api/utils.png" >
      
        <meta  property="og:image:type"  content="image/png" >
      
        <meta  property="og:image:width"  content="1200" >
      
        <meta  property="og:image:height"  content="630" >
      
        <meta  property="og:url"  content="None" >
      
        <meta  name="twitter:card"  content="summary_large_image" >
      
        <meta  name="twitter:title"  content="Utility - CONCORD - Contrastive Learning for Cross-domain Reconciliation and Discovery" >
      
        <meta  name="twitter:description"  content="None" >
      
        <meta  name="twitter:image"  content="./assets/images/social/api/utils.png" >
      
    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="purple">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#utilities" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow md-header--lifted" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../.." title="CONCORD - Contrastive Learning for Cross-domain Reconciliation and Discovery" class="md-header__button md-logo" aria-label="CONCORD - Contrastive Learning for Cross-domain Reconciliation and Discovery" data-md-component="logo">
      
  <img src="../../images/logo.png" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            CONCORD - Contrastive Learning for Cross-domain Reconciliation and Discovery
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Utility
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="purple"  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_0">
    
      <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M17 6H7c-3.31 0-6 2.69-6 6s2.69 6 6 6h10c3.31 0 6-2.69 6-6s-2.69-6-6-6m0 10H7c-2.21 0-4-1.79-4-4s1.79-4 4-4h10c2.21 0 4 1.79 4 4s-1.79 4-4 4M7 9c-1.66 0-3 1.34-3 3s1.34 3 3 3 3-1.34 3-3-1.34-3-3-3"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="" data-md-color-scheme="slate" data-md-color-primary="indigo" data-md-color-accent="lime"  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_1">
    
      <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_0" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M17 7H7a5 5 0 0 0-5 5 5 5 0 0 0 5 5h10a5 5 0 0 0 5-5 5 5 0 0 0-5-5m0 8a3 3 0 0 1-3-3 3 3 0 0 1 3-3 3 3 0 0 1 3 3 3 3 0 0 1-3 3"/></svg>
      </label>
    
  
</form>
      
    
    
      <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
        <div class="md-search__suggest" data-md-component="search-suggest"></div>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
  </nav>
  
    
      
<nav class="md-tabs" aria-label="Tabs" data-md-component="tabs">
  <div class="md-grid">
    <ul class="md-tabs__list">
      
        
  
  
  
    <li class="md-tabs__item">
      <a href="../.." class="md-tabs__link">
        
  
    
  
  Getting started

      </a>
    </li>
  

      
        
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../../galleries/cbce_show/" class="md-tabs__link">
          
  
  Galleries

        </a>
      </li>
    
  

      
        
  
  
  
    
    
      
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../../notebooks/concord_pbmc3k/" class="md-tabs__link">
          
  
  Tutorials

        </a>
      </li>
    
  

    
  

      
        
  
  
  
    
    
      
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../../notebooks/simulation_trajectory_show/" class="md-tabs__link">
          
  
  Advanced usage

        </a>
      </li>
    
  

    
  

      
        
  
  
    
  
  
    
    
      <li class="md-tabs__item md-tabs__item--active">
        <a href="../main/" class="md-tabs__link">
          
  
  API

        </a>
      </li>
    
  

      
        
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../../about/" class="md-tabs__link">
          
  
  About

        </a>
      </li>
    
  

      
    </ul>
  </div>
</nav>
    
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


  


  

<nav class="md-nav md-nav--primary md-nav--lifted md-nav--integrated" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../.." title="CONCORD - Contrastive Learning for Cross-domain Reconciliation and Discovery" class="md-nav__button md-logo" aria-label="CONCORD - Contrastive Learning for Cross-domain Reconciliation and Discovery" data-md-component="logo">
      
  <img src="../../images/logo.png" alt="logo">

    </a>
    CONCORD - Contrastive Learning for Cross-domain Reconciliation and Discovery
  </label>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Getting started
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_2" >
        
          
          <label class="md-nav__link" for="__nav_2" id="__nav_2_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Galleries
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_2">
            <span class="md-nav__icon md-icon"></span>
            Galleries
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../galleries/cbce_show/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    C.elegans/C.briggsae Embryogenesis Atlas
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../galleries/huycke_show/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Intestine development Atlas
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_3" >
        
          
          <label class="md-nav__link" for="__nav_3" id="__nav_3_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Tutorials
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3">
            <span class="md-nav__icon md-icon"></span>
            Tutorials
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_3_1" >
        
          
          <label class="md-nav__link" for="__nav_3_1" id="__nav_3_1_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Single batch, RNA-seq
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_3_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3_1">
            <span class="md-nav__icon md-icon"></span>
            Single batch, RNA-seq
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../notebooks/concord_pbmc3k/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    PBMC3k dataset, single batch
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_3_2" >
        
          
          <label class="md-nav__link" for="__nav_3_2" id="__nav_3_2_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Multiple batch, RNA-seq
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_3_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3_2">
            <span class="md-nav__icon md-icon"></span>
            Multiple batch, RNA-seq
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../notebooks/concord_Huycke/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Intestine development by Huycke et al.
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../notebooks/concord_mouse_organogenesis_100k/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Mouse organogenesis (100k subset for tutorial)
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_4" >
        
          
          <label class="md-nav__link" for="__nav_4" id="__nav_4_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Advanced usage
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_4_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4">
            <span class="md-nav__icon md-icon"></span>
            Advanced usage
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_4_1" >
        
          
          <label class="md-nav__link" for="__nav_4_1" id="__nav_4_1_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Simulation pipeline
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_4_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1">
            <span class="md-nav__icon md-icon"></span>
            Simulation pipeline
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../notebooks/simulation_trajectory_show/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Trajectory simulation
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
    
      
        
        
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5" checked>
        
          
          <label class="md-nav__link" for="__nav_5" id="__nav_5_label" tabindex="">
            
  
  <span class="md-ellipsis">
    API
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_5_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_5">
            <span class="md-nav__icon md-icon"></span>
            API
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../main/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Concord
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../model/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Model
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  <span class="md-ellipsis">
    Utility
    
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  <span class="md-ellipsis">
    Utility
    
  </span>
  

      </a>
      
        

  

<nav class="md-nav md-nav--secondary" aria-label="Table of Contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of Contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#Concord.utils.benchmark_topology" class="md-nav__link">
    <span class="md-ellipsis">
      benchmark_topology
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#Concord.utils.compute_persistent_homology" class="md-nav__link">
    <span class="md-ellipsis">
      compute_persistent_homology
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#Concord.utils.compute_betti_median_or_mode" class="md-nav__link">
    <span class="md-ellipsis">
      compute_betti_median_or_mode
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#Concord.utils.compute_betti_entropy" class="md-nav__link">
    <span class="md-ellipsis">
      compute_betti_entropy
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#Concord.utils.interpolate_betti_curve" class="md-nav__link">
    <span class="md-ellipsis">
      interpolate_betti_curve
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#Concord.utils.compute_betti_statistics" class="md-nav__link">
    <span class="md-ellipsis">
      compute_betti_statistics
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#Concord.utils.summarize_betti_statistics" class="md-nav__link">
    <span class="md-ellipsis">
      summarize_betti_statistics
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#Concord.utils.benchmark_geometry" class="md-nav__link">
    <span class="md-ellipsis">
      benchmark_geometry
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#Concord.utils.pairwise_distance" class="md-nav__link">
    <span class="md-ellipsis">
      pairwise_distance
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#Concord.utils.local_vs_distal_corr" class="md-nav__link">
    <span class="md-ellipsis">
      local_vs_distal_corr
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#Concord.utils.compute_state_batch_distance_ratio" class="md-nav__link">
    <span class="md-ellipsis">
      compute_state_batch_distance_ratio
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#Concord.utils.compute_trustworthiness" class="md-nav__link">
    <span class="md-ellipsis">
      compute_trustworthiness
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#Concord.utils.Simulation" class="md-nav__link">
    <span class="md-ellipsis">
      Simulation
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#Concord.utils.select_features" class="md-nav__link">
    <span class="md-ellipsis">
      select_features
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#Concord.utils.generate_synthetic_doublets" class="md-nav__link">
    <span class="md-ellipsis">
      generate_synthetic_doublets
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#Concord.utils.list_adata_files" class="md-nav__link">
    <span class="md-ellipsis">
      list_adata_files
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#Concord.utils.read_and_concatenate_adata" class="md-nav__link">
    <span class="md-ellipsis">
      read_and_concatenate_adata
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#Concord.utils.filter_and_copy_attributes" class="md-nav__link">
    <span class="md-ellipsis">
      filter_and_copy_attributes
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#Concord.utils.ensure_categorical" class="md-nav__link">
    <span class="md-ellipsis">
      ensure_categorical
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#Concord.utils.save_obsm_to_hdf5" class="md-nav__link">
    <span class="md-ellipsis">
      save_obsm_to_hdf5
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#Concord.utils.load_obsm_from_hdf5" class="md-nav__link">
    <span class="md-ellipsis">
      load_obsm_from_hdf5
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#Concord.utils.subset_adata_to_obsm_indices" class="md-nav__link">
    <span class="md-ellipsis">
      subset_adata_to_obsm_indices
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#Concord.utils.anndata_to_viscello" class="md-nav__link">
    <span class="md-ellipsis">
      anndata_to_viscello
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#Concord.utils.update_clist_with_subsets" class="md-nav__link">
    <span class="md-ellipsis">
      update_clist_with_subsets
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../plotting/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Plotting
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_6" >
        
          
          <label class="md-nav__link" for="__nav_6" id="__nav_6_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    About
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_6_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_6">
            <span class="md-nav__icon md-icon"></span>
            About
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../about/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Citation & Feedback
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../LICENSE/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    License
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  



<h1 id="utilities">Utilities</h1>
<p><code>utils</code> can be replaced by <code>ul</code>, e.g., <code>Concord.utils.list_adata_files</code> can be <code>Concord.ul.list_adata_files</code></p>


<div class="doc doc-object doc-function">


<h2 id="Concord.utils.benchmark_topology" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">Concord</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">benchmark_topology</span><span class="p">(</span><span class="n">diagrams</span><span class="p">,</span> <span class="n">expected_betti_numbers</span><span class="o">=</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span> <span class="n">n_bins</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">save_dir</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">file_suffix</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents first">

        <p>Benchmark the topological properties of persistence diagrams.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>diagrams</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>dict
A dictionary where keys are method names and values are persistence diagrams.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>expected_betti_numbers</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>list, optional
A list specifying the expected Betti numbers for different homology dimensions. Default is [1, 0, 0].</p>
              </div>
            </td>
            <td>
                  <code>[1, 0, 0]</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>n_bins</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>int, optional
Number of bins to use for Betti curve calculations. Default is 100.</p>
              </div>
            </td>
            <td>
                  <code>100</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>save_dir</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>str, optional
Directory to save benchmarking results as CSV files. If None, results are not saved.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>file_suffix</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>str, optional
Suffix to append to saved filenames.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>dict
A dictionary containing:
- <code>'betti_stats'</code>: DataFrame summarizing Betti statistics.
- <code>'distance_metrics'</code>: DataFrame of computed distances between Betti curves.
- <code>'combined_metrics'</code>: DataFrame of entropy, variance, and L1 distance metrics.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="Concord.utils.compute_persistent_homology" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">Concord</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">compute_persistent_homology</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="s1">&#39;X_pca&#39;</span><span class="p">,</span> <span class="n">homology_dimensions</span><span class="o">=</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">])</span></code>

</h2>


    <div class="doc doc-contents first">

        <p>Computes persistent homology using Vietoris-Rips complex.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>adata</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>anndata.AnnData
The AnnData object containing the data.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>key</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>str, optional
The key in <code>adata.obsm</code> specifying the embedding to use. Default is 'X_pca'.</p>
              </div>
            </td>
            <td>
                  <code>&#39;X_pca&#39;</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>homology_dimensions</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>list, optional
List of homology dimensions to compute. Default is [0, 1, 2].</p>
              </div>
            </td>
            <td>
                  <code>[0, 1, 2]</code>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>np.ndarray
Persistence diagrams representing homology classes across filtration values.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="Concord.utils.compute_betti_median_or_mode" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">Concord</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">compute_betti_median_or_mode</span><span class="p">(</span><span class="n">betti_values</span><span class="p">,</span> <span class="n">statistic</span><span class="o">=</span><span class="s1">&#39;median&#39;</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents first">

        <p>Computes the median or mode of Betti numbers.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>betti_values</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>np.ndarray
Array of Betti numbers across filtration values.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>statistic</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>str, optional
Statistic to compute ('median' or 'mode'). Default is 'median'.</p>
              </div>
            </td>
            <td>
                  <code>&#39;median&#39;</code>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>float
The computed median or mode of the Betti numbers.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="Concord.utils.compute_betti_entropy" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">Concord</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">compute_betti_entropy</span><span class="p">(</span><span class="n">betti_values</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents first">

        <p>Computes the entropy of the Betti curve.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>betti_values</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>np.ndarray
Array of Betti numbers across filtration values.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>float
The entropy of the Betti curve.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="Concord.utils.interpolate_betti_curve" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">Concord</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">interpolate_betti_curve</span><span class="p">(</span><span class="n">betti_values</span><span class="p">,</span> <span class="n">original_sampling</span><span class="p">,</span> <span class="n">common_sampling</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents first">

        <p>Interpolates Betti curve onto a common filtration grid.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>betti_values</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>np.ndarray
Array of Betti numbers.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>original_sampling</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>np.ndarray
The original filtration values associated with the Betti numbers.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>common_sampling</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>np.ndarray
The target filtration values for interpolation.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>np.ndarray
Interpolated Betti curve.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="Concord.utils.compute_betti_statistics" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">Concord</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">compute_betti_statistics</span><span class="p">(</span><span class="n">diagram</span><span class="p">,</span> <span class="n">expected_betti_numbers</span><span class="p">,</span> <span class="n">n_bins</span><span class="o">=</span><span class="mi">100</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents first">

        <p>Computes Betti statistics given a persistence diagram.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>diagram</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>np.ndarray
Persistence diagram from Giotto-TDA.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>expected_betti_numbers</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>np.ndarray
Expected Betti numbers for different homology dimensions.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>n_bins</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>int, optional
Number of bins for the Betti curve computation. Default is 100.</p>
              </div>
            </td>
            <td>
                  <code>100</code>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>dict
A dictionary containing:
- <code>'betti_stats'</code>: Dictionary of Betti statistics.
- <code>'observed_betti_numbers'</code>: Observed Betti numbers.
- <code>'expected_betti_numbers'</code>: Expected Betti numbers.
- <code>'l1_distance'</code>: L1 distance between observed and expected Betti numbers.
- <code>'l2_distance'</code>: L2 distance between observed and expected Betti numbers.
- <code>'total_relative_error'</code>: Total relative error.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="Concord.utils.summarize_betti_statistics" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">Concord</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">summarize_betti_statistics</span><span class="p">(</span><span class="n">betti_stats</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents first">

        <p>Summarizes Betti statistics into pandas DataFrames.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>betti_stats</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>dict
Dictionary containing Betti statistics for different methods.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>tuple
- <code>betti_stats_pivot</code>: DataFrame of Betti statistics.
- <code>distance_metrics_df</code>: DataFrame of distance metrics.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="Concord.utils.benchmark_geometry" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">Concord</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">benchmark_geometry</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">keys</span><span class="p">,</span> <span class="n">eval_metrics</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;pseudotime&#39;</span><span class="p">,</span> <span class="s1">&#39;cell_distance_corr&#39;</span><span class="p">,</span> <span class="s1">&#39;local_distal_corr&#39;</span><span class="p">,</span> <span class="s1">&#39;trustworthiness&#39;</span><span class="p">,</span> <span class="s1">&#39;state_distance_corr&#39;</span><span class="p">,</span> <span class="s1">&#39;state_dispersion_corr&#39;</span><span class="p">,</span> <span class="s1">&#39;state_batch_distance_ratio&#39;</span><span class="p">],</span> <span class="n">dist_metric</span><span class="o">=</span><span class="s1">&#39;cosine&#39;</span><span class="p">,</span> <span class="n">groundtruth_key</span><span class="o">=</span><span class="s1">&#39;PCA_no_noise&#39;</span><span class="p">,</span> <span class="n">state_key</span><span class="o">=</span><span class="s1">&#39;cluster&#39;</span><span class="p">,</span> <span class="n">batch_key</span><span class="o">=</span><span class="s1">&#39;batch&#39;</span><span class="p">,</span> <span class="n">groundtruth_dispersion</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">ground_truth_dispersion_key</span><span class="o">=</span><span class="s1">&#39;wt_noise&#39;</span><span class="p">,</span> <span class="n">corr_types</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;pearsonr&#39;</span><span class="p">,</span> <span class="s1">&#39;spearmanr&#39;</span><span class="p">,</span> <span class="s1">&#39;kendalltau&#39;</span><span class="p">],</span> <span class="n">trustworthiness_n_neighbors</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="mi">101</span><span class="p">,</span> <span class="mi">10</span><span class="p">),</span> <span class="n">dispersion_metric</span><span class="o">=</span><span class="s1">&#39;var&#39;</span><span class="p">,</span> <span class="n">return_type</span><span class="o">=</span><span class="s1">&#39;dataframe&#39;</span><span class="p">,</span> <span class="n">local_percentile</span><span class="o">=</span><span class="mf">0.1</span><span class="p">,</span> <span class="n">distal_percentile</span><span class="o">=</span><span class="mf">0.9</span><span class="p">,</span> <span class="n">start_point</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">end_point</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">pseudotime_k</span><span class="o">=</span><span class="mi">30</span><span class="p">,</span> <span class="n">truetime_key</span><span class="o">=</span><span class="s1">&#39;time&#39;</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">save_dir</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">file_suffix</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents first">

        <p>Benchmark the geometric properties of different embeddings.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>adata</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>anndata.AnnData
The AnnData object containing cell embeddings.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>keys</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>list
List of embeddings (keys in <code>adata.obsm</code>) to evaluate.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>eval_metrics</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>list, optional
Metrics to compute, such as 'pseudotime', 'cell_distance_corr', etc. Default includes multiple metrics.</p>
              </div>
            </td>
            <td>
                  <code>[&#39;pseudotime&#39;, &#39;cell_distance_corr&#39;, &#39;local_distal_corr&#39;, &#39;trustworthiness&#39;, &#39;state_distance_corr&#39;, &#39;state_dispersion_corr&#39;, &#39;state_batch_distance_ratio&#39;]</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>dist_metric</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>str, optional
Distance metric for computing cell distances. Default is 'cosine'.</p>
              </div>
            </td>
            <td>
                  <code>&#39;cosine&#39;</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>groundtruth_key</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>str, optional
Key in <code>adata.obsm</code> containing the ground truth embedding. Default is 'PCA_no_noise'.</p>
              </div>
            </td>
            <td>
                  <code>&#39;PCA_no_noise&#39;</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>state_key</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>str, optional
Key in <code>adata.obs</code> representing cell states or clusters.</p>
              </div>
            </td>
            <td>
                  <code>&#39;cluster&#39;</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>batch_key</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>str, optional
Key in <code>adata.obs</code> representing batch information.</p>
              </div>
            </td>
            <td>
                  <code>&#39;batch&#39;</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>groundtruth_dispersion</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>dict, optional
Precomputed dispersion values for ground truth, if available.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>ground_truth_dispersion_key</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>str, optional
Key used when computing dispersion correlations. Default is 'wt_noise'.</p>
              </div>
            </td>
            <td>
                  <code>&#39;wt_noise&#39;</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>corr_types</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>list, optional
List of correlation methods to compute. Default includes 'pearsonr', 'spearmanr', and 'kendalltau'.</p>
              </div>
            </td>
            <td>
                  <code>[&#39;pearsonr&#39;, &#39;spearmanr&#39;, &#39;kendalltau&#39;]</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>trustworthiness_n_neighbors</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>np.ndarray, optional
Range of neighborhood sizes for trustworthiness computation. Default is <code>np.arange(10, 101, 10)</code>.</p>
              </div>
            </td>
            <td>
                  <code><span title="numpy.arange">arange</span>(10, 101, 10)</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>dispersion_metric</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>str, optional
Metric to compute dispersion, e.g., 'var' (variance). Default is 'var'.</p>
              </div>
            </td>
            <td>
                  <code>&#39;var&#39;</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>return_type</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>str, optional
If 'dataframe', returns summary statistics; if 'full', returns additional details. Default is 'dataframe'.</p>
              </div>
            </td>
            <td>
                  <code>&#39;dataframe&#39;</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>local_percentile</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>float, optional
Percentile threshold for local distance correlations. Default is 0.1.</p>
              </div>
            </td>
            <td>
                  <code>0.1</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>distal_percentile</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>float, optional
Percentile threshold for distal distance correlations. Default is 0.9.</p>
              </div>
            </td>
            <td>
                  <code>0.9</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>start_point</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>int, optional
Index of the starting cell for pseudotime computation. Must be specified.</p>
              </div>
            </td>
            <td>
                  <code>0</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>end_point</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>int, optional
Index of the ending cell for pseudotime computation. Must be specified.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>pseudotime_k</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>int, optional
Number of neighbors used in k-NN graph for pseudotime computation. Default is 30.</p>
              </div>
            </td>
            <td>
                  <code>30</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>truetime_key</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>str, optional
Key in <code>adata.obs</code> representing ground truth time. Default is 'time'.</p>
              </div>
            </td>
            <td>
                  <code>&#39;time&#39;</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>verbose</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>bool, optional
Whether to enable logging. Default is True.</p>
              </div>
            </td>
            <td>
                  <code>True</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>save_dir</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>str, optional
Directory to save benchmarking results. If None, results are not saved.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>file_suffix</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>str, optional
Suffix for saved filenames.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>pd.DataFrame or tuple
If <code>return_type='dataframe'</code>, returns a DataFrame summarizing benchmark results.
If <code>return_type='full'</code>, returns both the DataFrame and a detailed results dictionary.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="Concord.utils.pairwise_distance" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">Concord</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">pairwise_distance</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">keys</span><span class="p">,</span> <span class="n">metric</span><span class="o">=</span><span class="s1">&#39;cosine&#39;</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents first">

    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="Concord.utils.local_vs_distal_corr" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">Concord</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">local_vs_distal_corr</span><span class="p">(</span><span class="n">X_high</span><span class="p">,</span> <span class="n">X_low</span><span class="p">,</span> <span class="n">local_percentile</span><span class="o">=</span><span class="mi">25</span><span class="p">,</span> <span class="n">distal_percentile</span><span class="o">=</span><span class="mi">75</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s1">&#39;pearsonr&#39;</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents first">

        <p>Computes correlation between local and distal pairwise distances.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>X_high</code>
            </td>
            <td>
                  <code><span title="numpy.ndarray">ndarray</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>High-dimensional data matrix.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>X_low</code>
            </td>
            <td>
                  <code><span title="numpy.ndarray">ndarray</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Low-dimensional embedding matrix.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>local_percentile</code>
            </td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Percentile threshold for local distances. Defaults to 25.</p>
              </div>
            </td>
            <td>
                  <code>25</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>distal_percentile</code>
            </td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Percentile threshold for distal distances. Defaults to 75.</p>
              </div>
            </td>
            <td>
                  <code>75</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>method</code>
            </td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Correlation method; 'pearsonr', 'spearmanr', or 'kendalltau'. Defaults to 'pearsonr'.</p>
              </div>
            </td>
            <td>
                  <code>&#39;pearsonr&#39;</code>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>float</code></td>            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Correlation for local distances.</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
<td><code>float</code></td>            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Correlation for distal distances.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="Concord.utils.compute_state_batch_distance_ratio" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">Concord</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">compute_state_batch_distance_ratio</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">basis</span><span class="o">=</span><span class="s1">&#39;X_latent&#39;</span><span class="p">,</span> <span class="n">batch_key</span><span class="o">=</span><span class="s1">&#39;batch&#39;</span><span class="p">,</span> <span class="n">state_key</span><span class="o">=</span><span class="s1">&#39;cluster&#39;</span><span class="p">,</span> <span class="n">metric</span><span class="o">=</span><span class="s1">&#39;cosine&#39;</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents first">

        <p>Computes the Batch-to-State Distance Ratio using centroids to evaluate batch correction.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>adata</code>
            </td>
            <td>
                  <code>AnnData</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>AnnData object containing latent embeddings.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>basis</code>
            </td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Key for latent embeddings in adata.obsm. Defaults to 'X_latent'.</p>
              </div>
            </td>
            <td>
                  <code>&#39;X_latent&#39;</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>batch_key</code>
            </td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Key for batch labels in adata.obs. Defaults to 'batch'.</p>
              </div>
            </td>
            <td>
                  <code>&#39;batch&#39;</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>state_key</code>
            </td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Key for cell state labels in adata.obs. Defaults to 'cluster'.</p>
              </div>
            </td>
            <td>
                  <code>&#39;cluster&#39;</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>metric</code>
            </td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Distance metric to use, e.g., 'cosine' or 'euclidean'. Defaults to 'cosine'.</p>
              </div>
            </td>
            <td>
                  <code>&#39;cosine&#39;</code>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>float</code></td>            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Ratio of average batch distance to average state distance.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="Concord.utils.compute_trustworthiness" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">Concord</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">compute_trustworthiness</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">embedding_keys</span><span class="p">,</span> <span class="n">groundtruth</span><span class="p">,</span> <span class="n">metric</span><span class="o">=</span><span class="s1">&#39;euclidean&#39;</span><span class="p">,</span> <span class="n">n_neighbors</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents first">

        <p>Evaluates trustworthiness of embeddings in an AnnData object.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>adata</code>
            </td>
            <td>
                  <code>AnnData</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>AnnData object containing embeddings in adata.obsm.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>embedding_keys</code>
            </td>
            <td>
                  <code>list</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>List of keys in adata.obsm to evaluate (e.g., ['X_umap', 'X_tsne']).</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>groundtruth</code>
            </td>
            <td>
                  <code>str or <span title="numpy.ndarray">ndarray</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Key in adata.obsm or adata.layers for ground truth data, or a precomputed matrix.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>metric</code>
            </td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Distance metric for trustworthiness calculation, e.g., 'euclidean' or 'cosine'. Defaults to 'euclidean'.</p>
              </div>
            </td>
            <td>
                  <code>&#39;euclidean&#39;</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>n_neighbors</code>
            </td>
            <td>
                  <code>int or list</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Neighborhood sizes for trustworthiness evaluation. Defaults to 10.</p>
              </div>
            </td>
            <td>
                  <code>10</code>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>pandas.DataFrame: Trustworthiness scores for each embedding at each neighborhood size.</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>pandas.DataFrame: Summary statistics with average trustworthiness and decay rate.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

    </div>

</div>

<div class="doc doc-object doc-class">



<h2 id="Concord.utils.Simulation" class="doc doc-heading">
            <code>Concord.utils.Simulation</code>


</h2>


    <div class="doc doc-contents first">


        <p>A class for simulating single-cell gene expression data with various structures and batch effects.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>n_cells</code>
            </td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Number of cells to simulate. Defaults to 1000.</p>
              </div>
            </td>
            <td>
                  <code>1000</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>n_genes</code>
            </td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Number of genes to simulate. Defaults to 1000.</p>
              </div>
            </td>
            <td>
                  <code>1000</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>n_batches</code>
            </td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Number of batches to simulate. Defaults to 2.</p>
              </div>
            </td>
            <td>
                  <code>2</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>n_states</code>
            </td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Number of states (e.g., clusters, trajectories). Defaults to 3.</p>
              </div>
            </td>
            <td>
                  <code>3</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>state_type</code>
            </td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Type of state to simulate; options include 'cluster', 'trajectory', 'tree', etc. Defaults to 'cluster'.</p>
              </div>
            </td>
            <td>
                  <code>&#39;cluster&#39;</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>batch_type</code>
            </td>
            <td>
                  <code>str or list</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Type of batch effect; options include 'batch_specific_features', 'variance_inflation', etc. Defaults to 'batch_specific_features'.</p>
              </div>
            </td>
            <td>
                  <code>&#39;batch_specific_features&#39;</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>state_distribution</code>
            </td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Distribution type for states; e.g., 'normal', 'poisson'. Defaults to 'normal'.</p>
              </div>
            </td>
            <td>
                  <code>&#39;normal&#39;</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>state_level</code>
            </td>
            <td>
                  <code>float</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Mean expression level for states. Defaults to 1.0.</p>
              </div>
            </td>
            <td>
                  <code>1.0</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>state_min_level</code>
            </td>
            <td>
                  <code>float</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Minimum expression level. Defaults to 0.0.</p>
              </div>
            </td>
            <td>
                  <code>0.0</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>state_dispersion</code>
            </td>
            <td>
                  <code>float</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Dispersion of state expression. Defaults to 0.1.</p>
              </div>
            </td>
            <td>
                  <code>0.1</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>program_structure</code>
            </td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Gene expression program structure; e.g., 'linear', 'bidirectional'. Defaults to "linear".</p>
              </div>
            </td>
            <td>
                  <code>&#39;linear&#39;</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>program_on_time_fraction</code>
            </td>
            <td>
                  <code>float</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Fraction of time the program is on. Defaults to 0.3.</p>
              </div>
            </td>
            <td>
                  <code>0.3</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>program_gap_size</code>
            </td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Size of gaps in expression programs. Defaults to 1.</p>
              </div>
            </td>
            <td>
                  <code>1</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>program_noise_in_block</code>
            </td>
            <td>
                  <code>bool</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Whether to add noise within each expression block. Defaults to True.</p>
              </div>
            </td>
            <td>
                  <code>True</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>trajectory_program_num</code>
            </td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Number of programs in a trajectory simulation. Defaults to 3.</p>
              </div>
            </td>
            <td>
                  <code>3</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>trajectory_cell_block_size_ratio</code>
            </td>
            <td>
                  <code>float</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Ratio of cell block sizes in a trajectory. Defaults to 0.3.</p>
              </div>
            </td>
            <td>
                  <code>0.3</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>trajectory_loop_to</code>
            </td>
            <td>
                  <code>int or list</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Loop connection in trajectory simulations. Defaults to None.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>tree_branching_factor</code>
            </td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Number of branches per tree level. Defaults to 2.</p>
              </div>
            </td>
            <td>
                  <code>2</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>tree_depth</code>
            </td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Depth of the simulated tree. Defaults to 3.</p>
              </div>
            </td>
            <td>
                  <code>3</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>tree_program_decay</code>
            </td>
            <td>
                  <code>float</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Decay factor for tree programs across branches. Defaults to 0.5.</p>
              </div>
            </td>
            <td>
                  <code>0.5</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>tree_cellcount_decay</code>
            </td>
            <td>
                  <code>float</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Decay factor for cell numbers across tree branches. Defaults to 1.0.</p>
              </div>
            </td>
            <td>
                  <code>1.0</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>batch_distribution</code>
            </td>
            <td>
                  <code>str or list</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Distribution for batch effects. Defaults to 'normal'.</p>
              </div>
            </td>
            <td>
                  <code>&#39;normal&#39;</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>batch_level</code>
            </td>
            <td>
                  <code>float or list</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Magnitude of batch effects. Defaults to 1.0.</p>
              </div>
            </td>
            <td>
                  <code>1.0</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>batch_dispersion</code>
            </td>
            <td>
                  <code>float or list</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Dispersion of batch effects. Defaults to 0.1.</p>
              </div>
            </td>
            <td>
                  <code>0.1</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>batch_cell_proportion</code>
            </td>
            <td>
                  <code>list</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Proportion of cells per batch. Defaults to None.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>batch_feature_frac</code>
            </td>
            <td>
                  <code>float or list</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Fraction of genes affected by batch effects. Defaults to 0.1.</p>
              </div>
            </td>
            <td>
                  <code>0.1</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>global_non_specific_gene_fraction</code>
            </td>
            <td>
                  <code>float</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Fraction of genes that are globally non-specific. Defaults to 0.1.</p>
              </div>
            </td>
            <td>
                  <code>0.1</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>pairwise_non_specific_gene_fraction</code>
            </td>
            <td>
                  <code>dict</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Pairwise-specific gene fraction between state pairs. Defaults to None.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>universal_gene_fraction</code>
            </td>
            <td>
                  <code>float</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Fraction of universal genes expressed across all cells. Defaults to 0.0.</p>
              </div>
            </td>
            <td>
                  <code>0.0</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>non_neg</code>
            </td>
            <td>
                  <code>bool</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Whether to enforce non-negative expression values. Defaults to False.</p>
              </div>
            </td>
            <td>
                  <code>False</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>to_int</code>
            </td>
            <td>
                  <code>bool</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Whether to convert expression values to integers. Defaults to False.</p>
              </div>
            </td>
            <td>
                  <code>False</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>seed</code>
            </td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Random seed for reproducibility. Defaults to 0.</p>
              </div>
            </td>
            <td>
                  <code>0</code>
            </td>
          </tr>
      </tbody>
    </table>


<p><span class="doc-section-title">Methods:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
            <tr class="doc-section-item">
              <td><code><a class="autorefs autorefs-internal" title="Concord.utils.Simulation.simulate_data" href="#Concord.utils.Simulation.simulate_data">simulate_data</a></code></td>
              <td>
                <div class="doc-md-description">
                  <p>Simulates gene expression data, including batch effects.</p>
                </div>
              </td>
            </tr>
            <tr class="doc-section-item">
              <td><code><a class="autorefs autorefs-internal" title="Concord.utils.Simulation.simulate_state" href="#Concord.utils.Simulation.simulate_state">simulate_state</a></code></td>
              <td>
                <div class="doc-md-description">
                  <p>Simulates cell state-specific gene expression patterns.</p>
                </div>
              </td>
            </tr>
            <tr class="doc-section-item">
              <td><code><a class="autorefs autorefs-internal" title="Concord.utils.Simulation.simulate_batch" href="#Concord.utils.Simulation.simulate_batch">simulate_batch</a></code></td>
              <td>
                <div class="doc-md-description">
                  <p>Simulates batch-specific effects on gene expression.</p>
                </div>
              </td>
            </tr>
            <tr class="doc-section-item">
              <td><code><a class="autorefs autorefs-internal" title="Concord.utils.Simulation.simulate_clusters" href="#Concord.utils.Simulation.simulate_clusters">simulate_clusters</a></code></td>
              <td>
                <div class="doc-md-description">
                  <p>Simulates gene expression in discrete clusters.</p>
                </div>
              </td>
            </tr>
            <tr class="doc-section-item">
              <td><code><a class="autorefs autorefs-internal" title="Concord.utils.Simulation.simulate_trajectory" href="#Concord.utils.Simulation.simulate_trajectory">simulate_trajectory</a></code></td>
              <td>
                <div class="doc-md-description">
                  <p>Simulates continuous gene expression trajectories.</p>
                </div>
              </td>
            </tr>
            <tr class="doc-section-item">
              <td><code><a class="autorefs autorefs-internal" title="Concord.utils.Simulation.simulate_tree" href="#Concord.utils.Simulation.simulate_tree">simulate_tree</a></code></td>
              <td>
                <div class="doc-md-description">
                  <p>Simulates hierarchical branching gene expression.</p>
                </div>
              </td>
            </tr>
            <tr class="doc-section-item">
              <td><code><span title="Concord.utils.Simulation.simulate_gatto">simulate_gatto</span></code></td>
              <td>
                <div class="doc-md-description">
                  <p>Simulates expression patterns similar to Gatto et al., 2023.</p>
                </div>
              </td>
            </tr>
            <tr class="doc-section-item">
              <td><code><span title="Concord.utils.Simulation.simulate_s_curve">simulate_s_curve</span></code></td>
              <td>
                <div class="doc-md-description">
                  <p>Simulates an S-curve structure in gene expression.</p>
                </div>
              </td>
            </tr>
            <tr class="doc-section-item">
              <td><code><span title="Concord.utils.Simulation.simulate_swiss_roll">simulate_swiss_roll</span></code></td>
              <td>
                <div class="doc-md-description">
                  <p>Simulates a Swiss roll structure with optional hole.</p>
                </div>
              </td>
            </tr>
            <tr class="doc-section-item">
              <td><code><span title="Concord.utils.Simulation.simulate_expression_block">simulate_expression_block</span></code></td>
              <td>
                <div class="doc-md-description">
                  <p>Generates structured gene expression within a cell population.</p>
                </div>
              </td>
            </tr>
            <tr class="doc-section-item">
              <td><code><a class="autorefs autorefs-internal" title="Concord.utils.Simulation.simulate_dropout" href="#Concord.utils.Simulation.simulate_dropout">simulate_dropout</a></code></td>
              <td>
                <div class="doc-md-description">
                  <p>Simulates dropout in gene expression data.</p>
                </div>
              </td>
            </tr>
            <tr class="doc-section-item">
              <td><code><a class="autorefs autorefs-internal" title="Concord.utils.Simulation.downsample_mtx_umi" href="#Concord.utils.Simulation.downsample_mtx_umi">downsample_mtx_umi</a></code></td>
              <td>
                <div class="doc-md-description">
                  <p>Performs UMI count downsampling.</p>
                </div>
              </td>
            </tr>
            <tr class="doc-section-item">
              <td><code><span title="Concord.utils.Simulation.simulate_distribution">simulate_distribution</span></code></td>
              <td>
                <div class="doc-md-description">
                  <p>Samples values from specified distributions.</p>
                </div>
              </td>
            </tr>
      </tbody>
    </table>









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="Concord.utils.Simulation.downsample_mtx_umi" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">downsample_mtx_umi</span><span class="p">(</span><span class="n">mtx</span><span class="p">,</span> <span class="n">ratio</span><span class="o">=</span><span class="mf">0.1</span><span class="p">,</span> <span class="n">seed</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-staticmethod"><code>staticmethod</code></small>
  </span>

</h3>


    <div class="doc doc-contents ">

        <p>Simulates downsampling of a gene expression matrix (UMI counts) by a given ratio.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>mtx</code>
            </td>
            <td>
                  <code><span title="numpy.ndarray">ndarray</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The input matrix where rows represent genes and columns represent cells.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>ratio</code>
            </td>
            <td>
                  <code>float</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The downsampling ratio (default 0.1).</p>
              </div>
            </td>
            <td>
                  <code>0.1</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>seed</code>
            </td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Random seed for reproducibility (default 1).</p>
              </div>
            </td>
            <td>
                  <code>1</code>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>numpy.ndarray: The downsampled matrix.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="Concord.utils.Simulation.rnegbin" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">rnegbin</span><span class="p">(</span><span class="n">mu</span><span class="p">,</span> <span class="n">theta</span><span class="p">,</span> <span class="n">size</span><span class="p">)</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-staticmethod"><code>staticmethod</code></small>
  </span>

</h3>


    <div class="doc doc-contents ">

        <p>Generate random numbers from a negative binomial distribution.</p>
<p>Parameters:
n: Number of random numbers to generate.
mu: Mean of the distribution.
theta: Dispersion parameter.</p>

    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="Concord.utils.Simulation.simulate_batch" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">simulate_batch</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">cell_indices</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">cell_proportion</span><span class="o">=</span><span class="mf">0.3</span><span class="p">,</span> <span class="n">batch_name</span><span class="o">=</span><span class="s1">&#39;batch_1&#39;</span><span class="p">,</span> <span class="n">effect_type</span><span class="o">=</span><span class="s1">&#39;batch_specific_features&#39;</span><span class="p">,</span> <span class="n">distribution</span><span class="o">=</span><span class="s1">&#39;normal&#39;</span><span class="p">,</span> <span class="n">level</span><span class="o">=</span><span class="mf">1.0</span><span class="p">,</span> <span class="n">dispersion</span><span class="o">=</span><span class="mf">0.1</span><span class="p">,</span> <span class="n">batch_feature_frac</span><span class="o">=</span><span class="mf">0.1</span><span class="p">,</span> <span class="n">seed</span><span class="o">=</span><span class="mi">42</span><span class="p">)</span></code>

</h3>


    <div class="doc doc-contents ">

        <p>Applies batch-specific effects to an existing simulated dataset.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>adata</code>
            </td>
            <td>
                  <code>AnnData</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Base dataset to apply batch effects.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>cell_indices</code>
            </td>
            <td>
                  <code>array - like</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Indices of cells to modify. Defaults to None.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>cell_proportion</code>
            </td>
            <td>
                  <code>float</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Proportion of cells affected. Defaults to 0.3.</p>
              </div>
            </td>
            <td>
                  <code>0.3</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>batch_name</code>
            </td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Name of the batch. Defaults to 'batch_1'.</p>
              </div>
            </td>
            <td>
                  <code>&#39;batch_1&#39;</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>effect_type</code>
            </td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Type of batch effect (e.g., 'batch_specific_features', 'variance_inflation'). Defaults to 'batch_specific_features'.</p>
              </div>
            </td>
            <td>
                  <code>&#39;batch_specific_features&#39;</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>distribution</code>
            </td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Distribution type for batch effects (e.g., 'normal'). Defaults to 'normal'.</p>
              </div>
            </td>
            <td>
                  <code>&#39;normal&#39;</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>level</code>
            </td>
            <td>
                  <code>float</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Effect level (e.g., scaling factor). Defaults to 1.0.</p>
              </div>
            </td>
            <td>
                  <code>1.0</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>dispersion</code>
            </td>
            <td>
                  <code>float</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Dispersion of batch effects. Defaults to 0.1.</p>
              </div>
            </td>
            <td>
                  <code>0.1</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>batch_feature_frac</code>
            </td>
            <td>
                  <code>float</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Fraction of genes affected by batch effects. Defaults to 0.1.</p>
              </div>
            </td>
            <td>
                  <code>0.1</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>seed</code>
            </td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Random seed for reproducibility. Defaults to 42.</p>
              </div>
            </td>
            <td>
                  <code>42</code>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>tuple</code></td>            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <ul>
<li>batch_adata (AnnData): Modified dataset with batch effects.</li>
<li>batch_adata_pre (AnnData): Dataset before applying batch effects.</li>
</ul>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="Concord.utils.Simulation.simulate_clusters" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">simulate_clusters</span><span class="p">(</span><span class="n">n_genes</span><span class="o">=</span><span class="mi">6</span><span class="p">,</span> <span class="n">n_cells</span><span class="o">=</span><span class="mi">12</span><span class="p">,</span> <span class="n">num_clusters</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">program_structure</span><span class="o">=</span><span class="s1">&#39;uniform&#39;</span><span class="p">,</span> <span class="n">program_on_time_fraction</span><span class="o">=</span><span class="mf">0.3</span><span class="p">,</span> <span class="n">distribution</span><span class="o">=</span><span class="s1">&#39;normal&#39;</span><span class="p">,</span> <span class="n">mean_expression</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">min_expression</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">dispersion</span><span class="o">=</span><span class="mf">1.0</span><span class="p">,</span> <span class="n">global_non_specific_gene_fraction</span><span class="o">=</span><span class="mf">0.1</span><span class="p">,</span> <span class="n">pairwise_non_specific_gene_fraction</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">cluster_key</span><span class="o">=</span><span class="s1">&#39;cluster&#39;</span><span class="p">,</span> <span class="n">permute</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">seed</span><span class="o">=</span><span class="mi">42</span><span class="p">)</span></code>

</h3>


    <div class="doc doc-contents ">

        <p>Simulates gene expression for discrete cell clusters.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>n_genes</code>
            </td>
            <td>
                  <code>int or list</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Number of genes per cluster or total genes. Defaults to 6.</p>
              </div>
            </td>
            <td>
                  <code>6</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>n_cells</code>
            </td>
            <td>
                  <code>int or list</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Number of cells per cluster or total cells. Defaults to 12.</p>
              </div>
            </td>
            <td>
                  <code>12</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>num_clusters</code>
            </td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Number of clusters to simulate. Defaults to 2.</p>
              </div>
            </td>
            <td>
                  <code>2</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>program_structure</code>
            </td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Expression program structure ('linear', 'uniform', etc.). Defaults to 'uniform'.</p>
              </div>
            </td>
            <td>
                  <code>&#39;uniform&#39;</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>program_on_time_fraction</code>
            </td>
            <td>
                  <code>float</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Fraction of program duration. Defaults to 0.3.</p>
              </div>
            </td>
            <td>
                  <code>0.3</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>distribution</code>
            </td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Type of distribution for gene expression. Defaults to 'normal'.</p>
              </div>
            </td>
            <td>
                  <code>&#39;normal&#39;</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>mean_expression</code>
            </td>
            <td>
                  <code>float</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Mean expression level. Defaults to 10.</p>
              </div>
            </td>
            <td>
                  <code>10</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>min_expression</code>
            </td>
            <td>
                  <code>float</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Minimum expression level. Defaults to 1.</p>
              </div>
            </td>
            <td>
                  <code>1</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>dispersion</code>
            </td>
            <td>
                  <code>float</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Dispersion in expression levels. Defaults to 1.0.</p>
              </div>
            </td>
            <td>
                  <code>1.0</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>global_non_specific_gene_fraction</code>
            </td>
            <td>
                  <code>float</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Fraction of globally expressed genes. Defaults to 0.1.</p>
              </div>
            </td>
            <td>
                  <code>0.1</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>pairwise_non_specific_gene_fraction</code>
            </td>
            <td>
                  <code>dict</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Pairwise-specific genes between cluster pairs. Defaults to None.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>cluster_key</code>
            </td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Key for cluster labeling. Defaults to 'cluster'.</p>
              </div>
            </td>
            <td>
                  <code>&#39;cluster&#39;</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>permute</code>
            </td>
            <td>
                  <code>bool</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Whether to shuffle cells. Defaults to False.</p>
              </div>
            </td>
            <td>
                  <code>False</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>seed</code>
            </td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Random seed. Defaults to 42.</p>
              </div>
            </td>
            <td>
                  <code>42</code>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>AnnData</code></td>            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Simulated dataset with clustered gene expression.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="Concord.utils.Simulation.simulate_data" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">simulate_data</span><span class="p">()</span></code>

</h3>


    <div class="doc doc-contents ">

        <p>Simulates single-cell gene expression data, integrating state-based and batch effects.</p>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>tuple</code></td>            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <ul>
<li>adata (AnnData): Simulated gene expression data with batch effects.</li>
<li>adata_pre (AnnData): Pre-batch effect simulated data.</li>
</ul>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="Concord.utils.Simulation.simulate_dropout" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">simulate_dropout</span><span class="p">(</span><span class="n">mtx</span><span class="p">,</span> <span class="n">dropout_lambda</span><span class="o">=</span><span class="mf">1.0</span><span class="p">,</span> <span class="n">seed</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-staticmethod"><code>staticmethod</code></small>
  </span>

</h3>


    <div class="doc doc-contents ">

        <p>Simulates dropout in UMI counts based on the specified dropout lambda.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>mtx</code>
            </td>
            <td>
                  <code><span title="numpy.ndarray">ndarray</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The actual UMI counts matrix (genes x cells).</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>dropout_lambda</code>
            </td>
            <td>
                  <code>float</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The lambda parameter controlling the dropout probability.</p>
              </div>
            </td>
            <td>
                  <code>1.0</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>seed</code>
            </td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Seed for the random number generator for reproducibility.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>numpy.ndarray: The UMI counts matrix after applying dropout.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="Concord.utils.Simulation.simulate_state" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">simulate_state</span><span class="p">()</span></code>

</h3>


    <div class="doc doc-contents ">

        <p>Simulates gene expression profiles for different cell states.</p>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>AnnData</code></td>            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>An AnnData object containing simulated state-specific expression data.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="Concord.utils.Simulation.simulate_trajectory" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">simulate_trajectory</span><span class="p">(</span><span class="n">n_genes</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">n_cells</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">cell_block_size_ratio</span><span class="o">=</span><span class="mf">0.3</span><span class="p">,</span> <span class="n">program_num</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">program_structure</span><span class="o">=</span><span class="s1">&#39;linear&#39;</span><span class="p">,</span> <span class="n">program_on_time_fraction</span><span class="o">=</span><span class="mf">0.3</span><span class="p">,</span> <span class="n">distribution</span><span class="o">=</span><span class="s1">&#39;normal&#39;</span><span class="p">,</span> <span class="n">mean_expression</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">min_expression</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">dispersion</span><span class="o">=</span><span class="mf">1.0</span><span class="p">,</span> <span class="n">seed</span><span class="o">=</span><span class="mi">42</span><span class="p">,</span> <span class="n">loop_to</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>

</h3>


    <div class="doc doc-contents ">

        <p>Simulates a continuous trajectory of gene expression.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>n_genes</code>
            </td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Number of genes. Defaults to 10.</p>
              </div>
            </td>
            <td>
                  <code>10</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>n_cells</code>
            </td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Number of cells. Defaults to 100.</p>
              </div>
            </td>
            <td>
                  <code>100</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>cell_block_size_ratio</code>
            </td>
            <td>
                  <code>float</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Ratio of cell blocks. Defaults to 0.3.</p>
              </div>
            </td>
            <td>
                  <code>0.3</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>program_num</code>
            </td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Number of gene programs in the trajectory. Defaults to 3.</p>
              </div>
            </td>
            <td>
                  <code>3</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>program_structure</code>
            </td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Structure of gene programs ('linear', 'bidirectional'). Defaults to 'linear'.</p>
              </div>
            </td>
            <td>
                  <code>&#39;linear&#39;</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>program_on_time_fraction</code>
            </td>
            <td>
                  <code>float</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Fraction of time the program is on. Defaults to 0.3.</p>
              </div>
            </td>
            <td>
                  <code>0.3</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>distribution</code>
            </td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Distribution type. Defaults to 'normal'.</p>
              </div>
            </td>
            <td>
                  <code>&#39;normal&#39;</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>mean_expression</code>
            </td>
            <td>
                  <code>float</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Mean expression level. Defaults to 10.</p>
              </div>
            </td>
            <td>
                  <code>10</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>min_expression</code>
            </td>
            <td>
                  <code>float</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Minimum expression level. Defaults to 0.</p>
              </div>
            </td>
            <td>
                  <code>0</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>dispersion</code>
            </td>
            <td>
                  <code>float</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Dispersion of expression. Defaults to 1.0.</p>
              </div>
            </td>
            <td>
                  <code>1.0</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>seed</code>
            </td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Random seed. Defaults to 42.</p>
              </div>
            </td>
            <td>
                  <code>42</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>loop_to</code>
            </td>
            <td>
                  <code>int or list</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Defines looping relationships in the trajectory. Defaults to None.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>AnnData</code></td>            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Simulated dataset with continuous gene expression patterns.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="Concord.utils.Simulation.simulate_tree" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">simulate_tree</span><span class="p">(</span><span class="n">n_genes</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">n_cells</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">branching_factor</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">depth</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">program_structure</span><span class="o">=</span><span class="s1">&#39;linear_increasing&#39;</span><span class="p">,</span> <span class="n">program_on_time_fraction</span><span class="o">=</span><span class="mf">0.3</span><span class="p">,</span> <span class="n">program_gap_size</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">program_decay</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span> <span class="n">cellcount_decay</span><span class="o">=</span><span class="mf">1.0</span><span class="p">,</span> <span class="n">distribution</span><span class="o">=</span><span class="s1">&#39;normal&#39;</span><span class="p">,</span> <span class="n">mean_expression</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">min_expression</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">dispersion</span><span class="o">=</span><span class="mf">1.0</span><span class="p">,</span> <span class="n">seed</span><span class="o">=</span><span class="mi">42</span><span class="p">,</span> <span class="n">noise_in_block</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span></code>

</h3>


    <div class="doc doc-contents ">

        <p>Simulates hierarchical branching gene expression patterns.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>n_genes</code>
            </td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Number of genes. Defaults to 10.</p>
              </div>
            </td>
            <td>
                  <code>10</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>n_cells</code>
            </td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Number of cells. Defaults to 100.</p>
              </div>
            </td>
            <td>
                  <code>100</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>branching_factor</code>
            </td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Number of branches per level. Defaults to 2.</p>
              </div>
            </td>
            <td>
                  <code>2</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>depth</code>
            </td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Depth of the branching tree. Defaults to 3.</p>
              </div>
            </td>
            <td>
                  <code>3</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>program_structure</code>
            </td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Gene program structure. Defaults to 'linear_increasing'.</p>
              </div>
            </td>
            <td>
                  <code>&#39;linear_increasing&#39;</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>program_on_time_fraction</code>
            </td>
            <td>
                  <code>float</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Program activation time fraction. Defaults to 0.3.</p>
              </div>
            </td>
            <td>
                  <code>0.3</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>program_gap_size</code>
            </td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Gap size between programs. Defaults to 1.</p>
              </div>
            </td>
            <td>
                  <code>1</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>program_decay</code>
            </td>
            <td>
                  <code>float</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Decay factor for program effects. Defaults to 0.5.</p>
              </div>
            </td>
            <td>
                  <code>0.5</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>cellcount_decay</code>
            </td>
            <td>
                  <code>float</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Decay factor for cell counts. Defaults to 1.0.</p>
              </div>
            </td>
            <td>
                  <code>1.0</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>distribution</code>
            </td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Expression distribution type. Defaults to 'normal'.</p>
              </div>
            </td>
            <td>
                  <code>&#39;normal&#39;</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>mean_expression</code>
            </td>
            <td>
                  <code>float</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Mean gene expression level. Defaults to 10.</p>
              </div>
            </td>
            <td>
                  <code>10</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>min_expression</code>
            </td>
            <td>
                  <code>float</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Minimum gene expression level. Defaults to 0.</p>
              </div>
            </td>
            <td>
                  <code>0</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>dispersion</code>
            </td>
            <td>
                  <code>float</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Dispersion of expression. Defaults to 1.0.</p>
              </div>
            </td>
            <td>
                  <code>1.0</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>seed</code>
            </td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Random seed. Defaults to 42.</p>
              </div>
            </td>
            <td>
                  <code>42</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>noise_in_block</code>
            </td>
            <td>
                  <code>bool</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Whether to add noise within expression blocks. Defaults to True.</p>
              </div>
            </td>
            <td>
                  <code>True</code>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>AnnData</code></td>            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Simulated dataset with hierarchical tree-like gene expression.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="Concord.utils.select_features" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">Concord</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">select_features</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">n_top_features</span><span class="o">=</span><span class="mi">2000</span><span class="p">,</span> <span class="n">flavor</span><span class="o">=</span><span class="s1">&#39;seurat_v3&#39;</span><span class="p">,</span> <span class="n">filter_gene_by_counts</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">normalize</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">log1p</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">grouping</span><span class="o">=</span><span class="s1">&#39;cluster&#39;</span><span class="p">,</span> <span class="n">emb_key</span><span class="o">=</span><span class="s1">&#39;X_pca&#39;</span><span class="p">,</span> <span class="n">k</span><span class="o">=</span><span class="mi">512</span><span class="p">,</span> <span class="n">knn_samples</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">gini_cut_qt</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">save_path</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="mi">3</span><span class="p">),</span> <span class="n">subsample_frac</span><span class="o">=</span><span class="mf">1.0</span><span class="p">,</span> <span class="n">random_state</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents first">

        <p>Selects top informative features from an AnnData object.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>adata</code>
            </td>
            <td>
                  <code>AnnData</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>AnnData object containing gene expression data.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>n_top_features</code>
            </td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Number of top features to select. Defaults to 2000.</p>
              </div>
            </td>
            <td>
                  <code>2000</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>flavor</code>
            </td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Feature selection method. Options:
- 'seurat_v3': Highly variable gene selection based on Seurat v3.
- 'iff': Uses Informative Feature Filtering (IFF) method.
Defaults to "seurat_v3".</p>
              </div>
            </td>
            <td>
                  <code>&#39;seurat_v3&#39;</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>filter_gene_by_counts</code>
            </td>
            <td>
                  <code><span title="typing.Union">Union</span>[int, bool]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Minimum count threshold for feature filtering. Defaults to False.</p>
              </div>
            </td>
            <td>
                  <code>False</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>normalize</code>
            </td>
            <td>
                  <code>bool</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Whether to normalize the data before feature selection. Defaults to False.</p>
              </div>
            </td>
            <td>
                  <code>False</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>log1p</code>
            </td>
            <td>
                  <code>bool</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Whether to apply log1p transformation before feature selection. Defaults to False.</p>
              </div>
            </td>
            <td>
                  <code>False</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>grouping</code>
            </td>
            <td>
                  <code><span title="typing.Union">Union</span>[str, <span title="pandas.Series">Series</span>, <span title="typing.List">List</span>[str]]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Clustering/grouping strategy for IFF method. Defaults to 'cluster'.</p>
              </div>
            </td>
            <td>
                  <code>&#39;cluster&#39;</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>emb_key</code>
            </td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Embedding key in <code>adata.obsm</code> used for clustering. Defaults to 'X_pca'.</p>
              </div>
            </td>
            <td>
                  <code>&#39;X_pca&#39;</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>k</code>
            </td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Number of neighbors for k-NN if <code>grouping='knn'</code>. Defaults to 512.</p>
              </div>
            </td>
            <td>
                  <code>512</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>knn_samples</code>
            </td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Number of k-NN samples if <code>grouping='knn'</code>. Defaults to 100.</p>
              </div>
            </td>
            <td>
                  <code>100</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>gini_cut_qt</code>
            </td>
            <td>
                  <code>float</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Quantile threshold for selecting features by Gini coefficient in IFF. Defaults to None.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>save_path</code>
            </td>
            <td>
                  <code><span title="typing.Optional">Optional</span>[<span title="typing.Union">Union</span>[str, <span title="pathlib.Path">Path</span>]]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Path to save Gini coefficient plot. Defaults to None.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>figsize</code>
            </td>
            <td>
                  <code>tuple</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Size of Gini coefficient plot. Defaults to (10, 3).</p>
              </div>
            </td>
            <td>
                  <code>(10, 3)</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>subsample_frac</code>
            </td>
            <td>
                  <code>float</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Fraction of data to subsample for feature selection. Defaults to 1.0.</p>
              </div>
            </td>
            <td>
                  <code>1.0</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>random_state</code>
            </td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Random seed for reproducibility. Defaults to 0.</p>
              </div>
            </td>
            <td>
                  <code>0</code>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="typing.List">List</span>[str]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>List[str]: List of selected feature names.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="Concord.utils.generate_synthetic_doublets" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">Concord</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">generate_synthetic_doublets</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">doublet_synth_ratio</span><span class="p">,</span> <span class="n">seed</span><span class="p">,</span> <span class="n">batch_key</span><span class="p">,</span> <span class="n">droplet_type_key</span><span class="p">,</span> <span class="n">mean</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span> <span class="n">var</span><span class="o">=</span><span class="mf">0.1</span><span class="p">,</span> <span class="n">clip_range</span><span class="o">=</span><span class="p">(</span><span class="mf">0.2</span><span class="p">,</span> <span class="mf">0.8</span><span class="p">),</span> <span class="n">plot_histogram</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">combine_with_original</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents first">

        <p>Generate synthetic doublets from singlet data in an AnnData object within each batch.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>adata</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>AnnData object containing the singlet data (with maybe unclassified doublets)</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>doublet_synth_ratio</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>float, the ratio of synthetic doublets to true singlets</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>seed</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>int, random seed for reproducibility</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>batch_key</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>str, the key in .obs indicating batch information</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>droplet_type_key</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>str, the key in .obs indicating droplet type</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>mean</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>float, mean of the normal distribution for generating fractions (default: 0.5)</p>
              </div>
            </td>
            <td>
                  <code>0.5</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>var</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>float, variance of the normal distribution for generating fractions (default: 0.1)</p>
              </div>
            </td>
            <td>
                  <code>0.1</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>clip_range</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>tuple, range to clip the generated fractions (default: (0.2, 0.8))</p>
              </div>
            </td>
            <td>
                  <code>(0.2, 0.8)</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>plot_histogram</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>bool, whether to plot the histogram of synthetic doublet fractions</p>
              </div>
            </td>
            <td>
                  <code>True</code>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>adata_synthetic_doublets</code></td>            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>AnnData object containing the synthetic doublets</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="Concord.utils.list_adata_files" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">Concord</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">list_adata_files</span><span class="p">(</span><span class="n">folder_path</span><span class="p">,</span> <span class="n">substring</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">extension</span><span class="o">=</span><span class="s1">&#39;*.h5ad&#39;</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents first">

        <p>List all <code>.h5ad</code> files in a directory (recursively) that match a given substring.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>folder_path</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>str
Path to the folder where <code>.h5ad</code> files are located.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>substring</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>str, optional
A substring to filter filenames (default is None, meaning no filtering).</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>extension</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>str, optional
File extension to search for (default is "*.h5ad").</p>
              </div>
            </td>
            <td>
                  <code>&#39;*.h5ad&#39;</code>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>list
A list of file paths matching the criteria.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="Concord.utils.read_and_concatenate_adata" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">Concord</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">read_and_concatenate_adata</span><span class="p">(</span><span class="n">adata_files</span><span class="p">,</span> <span class="n">merge</span><span class="o">=</span><span class="s1">&#39;unique&#39;</span><span class="p">,</span> <span class="n">add_dataset_col</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">dataset_col_name</span><span class="o">=</span><span class="s1">&#39;dataset&#39;</span><span class="p">,</span> <span class="n">output_file</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents first">

        <p>Read and concatenate multiple AnnData <code>.h5ad</code> files into a single AnnData object.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>adata_files</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>list
List of file paths to <code>.h5ad</code> files to be concatenated.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>merge</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>str, optional
How to handle conflicting columns, e.g., 'unique' (default), 'first', etc.</p>
              </div>
            </td>
            <td>
                  <code>&#39;unique&#39;</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>add_dataset_col</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>bool, optional
Whether to add a new column in <code>adata.obs</code> identifying the source dataset.</p>
              </div>
            </td>
            <td>
                  <code>False</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>dataset_col_name</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>str, optional
Name of the new column storing dataset names.</p>
              </div>
            </td>
            <td>
                  <code>&#39;dataset&#39;</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>output_file</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>str, optional
Path to save the concatenated AnnData object. If None, the object is not saved.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>ad.AnnData
The concatenated AnnData object.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="Concord.utils.filter_and_copy_attributes" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">Concord</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">filter_and_copy_attributes</span><span class="p">(</span><span class="n">adata_target</span><span class="p">,</span> <span class="n">adata_source</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents first">

        <p>Filter <code>adata_target</code> to match the cells in <code>adata_source</code>, then copy <code>.obs</code> and <code>.obsm</code>.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>adata_target</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>ad.AnnData
The AnnData object to be filtered.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>adata_source</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>ad.AnnData
The reference AnnData object containing the desired cells and attributes.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>ad.AnnData
The filtered AnnData object with updated <code>.obs</code> and <code>.obsm</code>.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="Concord.utils.ensure_categorical" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">Concord</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">ensure_categorical</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">obs_key</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">drop_unused</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents first">

        <p>Convert an <code>.obs</code> column to categorical dtype.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>adata</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>ad.AnnData
The AnnData object.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>obs_key</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>str
Column in <code>.obs</code> to be converted to categorical.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>drop_unused</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>bool, optional
Whether to remove unused categories (default is True).</p>
              </div>
            </td>
            <td>
                  <code>True</code>
            </td>
          </tr>
      </tbody>
    </table>

    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="Concord.utils.save_obsm_to_hdf5" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">Concord</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">save_obsm_to_hdf5</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">filename</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents first">

        <p>Save the <code>.obsm</code> attribute of an AnnData object to an HDF5 file.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>adata</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>anndata.AnnData
The AnnData object containing the <code>.obsm</code> attribute to be saved.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>filename</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>str
The path to the HDF5 file where <code>.obsm</code> data will be stored.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>None
Saves <code>.obsm</code> data to the specified HDF5 file.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="Concord.utils.load_obsm_from_hdf5" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">Concord</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">load_obsm_from_hdf5</span><span class="p">(</span><span class="n">filename</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents first">

        <p>Load the <code>.obsm</code> attribute from an HDF5 file.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>filename</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>str
Path to the HDF5 file containing <code>.obsm</code> data.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>dict
A dictionary where keys are <code>.obsm</code> names and values are corresponding matrices.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="Concord.utils.subset_adata_to_obsm_indices" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">Concord</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">subset_adata_to_obsm_indices</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">obsm</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents first">

        <p>Subset an AnnData object to match the indices present in <code>.obsm</code>.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>adata</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>anndata.AnnData
The original AnnData object.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>obsm</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>dict
A dictionary containing <code>.obsm</code> data, where keys are embedding names, and values are arrays.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>anndata.AnnData
A subsetted AnnData object that contains only the indices available in <code>.obsm</code>.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="Concord.utils.anndata_to_viscello" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">Concord</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">anndata_to_viscello</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">output_dir</span><span class="p">,</span> <span class="n">project_name</span><span class="o">=</span><span class="s1">&#39;MyProject&#39;</span><span class="p">,</span> <span class="n">organism</span><span class="o">=</span><span class="s1">&#39;hsa&#39;</span><span class="p">,</span> <span class="n">clist_only</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents first">

        <p>Converts an AnnData object to a VisCello project directory.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>adata</code>
            </td>
            <td>
                  <code>AnnData</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>AnnData object containing single-cell data.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>output_dir</code>
            </td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Directory where the VisCello project will be created.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>project_name</code>
            </td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Name of the project. Defaults to "MyProject".</p>
              </div>
            </td>
            <td>
                  <code>&#39;MyProject&#39;</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>organism</code>
            </td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Organism code (e.g., 'hsa' for human). Defaults to 'hsa'.</p>
              </div>
            </td>
            <td>
                  <code>&#39;hsa&#39;</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>clist_only</code>
            </td>
            <td>
                  <code>bool</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Whether to generate only the clist file. Defaults to False.</p>
              </div>
            </td>
            <td>
                  <code>False</code>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>None</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


<details class="side-effects" open>
  <summary>Side Effects</summary>
  <ul>
<li>Creates a directory with the necessary files for VisCello.</li>
<li>Saves <code>eset.rds</code> (ExpressionSet), <code>config.yml</code>, and <code>clist.rds</code>.</li>
</ul>
</details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="Concord.utils.update_clist_with_subsets" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">Concord</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">update_clist_with_subsets</span><span class="p">(</span><span class="n">global_adata</span><span class="p">,</span> <span class="n">adata_subsets</span><span class="p">,</span> <span class="n">viscello_dir</span><span class="p">,</span> <span class="n">cluster_key</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents first">

        <p>Updates an existing VisCello clist with new subsets.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>global_adata</code>
            </td>
            <td>
                  <code>AnnData</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The full AnnData object.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>adata_subsets</code>
            </td>
            <td>
                  <code>dict</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Dictionary mapping subset names to AnnData objects.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>viscello_dir</code>
            </td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Path to the existing VisCello directory.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>cluster_key</code>
            </td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Key in <code>adata.obs</code> for cluster assignments. Defaults to None.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>None</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


<details class="side-effects" open>
  <summary>Side Effects</summary>
  <ul>
<li>Reads the existing <code>clist.rds</code> file from <code>viscello_dir</code>.</li>
<li>Adds new subsets as <code>Cello</code> objects to the clist.</li>
<li>Saves the updated <code>clist.rds</code> file in <code>viscello_dir</code>.</li>
</ul>
</details>
    </div>

</div>







  
  






                
              </article>
            </div>
          
          
  <script>var tabs=__md_get("__tabs");if(Array.isArray(tabs))e:for(var set of document.querySelectorAll(".tabbed-set")){var labels=set.querySelector(".tabbed-labels");for(var tab of tabs)for(var label of labels.getElementsByTagName("label"))if(label.innerText.trim()===tab){var input=document.getElementById(label.htmlFor);input.checked=!0;continue e}}</script>

<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
          <button type="button" class="md-top md-icon" data-md-component="top" hidden>
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8z"/></svg>
  Back to top
</button>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
    <div class="md-copyright__highlight">
      &copy; 2025 <a href="https://github.com/Gartner-Lab"  target="_blank" rel="noopener">Gartner Lab</a>
    </div>
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
        <div class="md-social">
  
    
    
    
    
      
      
    
    <a href="https://github.com/Gartner-Lab/Concord" target="_blank" rel="noopener" title="github.com" class="md-social__link">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 480 512"><!--! Font Awesome Free 6.7.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M186.1 328.7c0 20.9-10.9 55.1-36.7 55.1s-36.7-34.2-36.7-55.1 10.9-55.1 36.7-55.1 36.7 34.2 36.7 55.1M480 278.2c0 31.9-3.2 65.7-17.5 95-37.9 76.6-142.1 74.8-216.7 74.8-75.8 0-186.2 2.7-225.6-74.8-14.6-29-20.2-63.1-20.2-95 0-41.9 13.9-81.5 41.5-113.6-5.2-15.8-7.7-32.4-7.7-48.8 0-21.5 4.9-32.3 14.6-51.8 45.3 0 74.3 9 108.8 36 29-6.9 58.8-10 88.7-10 27 0 54.2 2.9 80.4 9.2 34-26.7 63-35.2 107.8-35.2 9.8 19.5 14.6 30.3 14.6 51.8 0 16.4-2.6 32.7-7.7 48.2 27.5 32.4 39 72.3 39 114.2m-64.3 50.5c0-43.9-26.7-82.6-73.5-82.6-18.9 0-37 3.4-56 6-14.9 2.3-29.8 3.2-45.1 3.2-15.2 0-30.1-.9-45.1-3.2-18.7-2.6-37-6-56-6-46.8 0-73.5 38.7-73.5 82.6 0 87.8 80.4 101.3 150.4 101.3h48.2c70.3 0 150.6-13.4 150.6-101.3m-82.6-55.1c-25.8 0-36.7 34.2-36.7 55.1s10.9 55.1 36.7 55.1 36.7-34.2 36.7-55.1-10.9-55.1-36.7-55.1"/></svg>
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    <script id="__config" type="application/json">{"base": "../..", "features": ["navigation.tabs", "navigation.tabs.sticky", "navigation.sections", "toc.integrate", "navigation.top", "navigation.sticky", "search.suggest", "search.highlight", "content.tabs.link", "content.code.annotation", "content.code.copy", "navigation.expand", "navigation.mobile", "navigation.instant", "content.full_width"], "search": "../../assets/javascripts/workers/search.f8cc74c7.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
    
    
      <script src="../../assets/javascripts/bundle.f1b6f286.min.js"></script>
      
        <script src="https://cdn.plot.ly/plotly-2.35.2.min.js"></script>
      
    
  </body>
</html>